File: amazon/909_snakes_and_ladders.java
import java.util.*;

class Solution {
    public int snakesAndLadders(int[][] board) {
        int n = board.length;
        
        // Call BFS to find the minimum number of moves
        return bfs(board, n);
    }

    private int bfs(int[][] board, int n) {
        boolean[] visited = new boolean[n * n + 1];
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{1, 0});  // Start at square 1 with 0 moves
        visited[1] = true;

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int currSquare = current[0];
            int moves = current[1];

            for (int i = 1; i <= 6; i++) {
                int nextSquare = currSquare + i;
                if (nextSquare > n * n) break;

                int[] pos = getPosition(nextSquare, n);
                int r = pos[0];
                int c = pos[1];

                if (board[r][c] != -1) {
                    nextSquare = board[r][c];  // Move to the destination of the snake/ladder
                }

                if (nextSquare == n * n) return moves + 1;  // Reached the last square

                if (!visited[nextSquare]) {
                    visited[nextSquare] = true;
                    queue.offer(new int[]{nextSquare, moves + 1});
                }
            }
        }

        return -1;  // If the last square is unreachable
    }

    private int[] getPosition(int square, int n) {
        int row = (square - 1) / n;
        int col = (square - 1) % n;
        if (row % 2 == 1) {
            col = n - 1 - col;  // Reverse column index on odd rows
        }
        return new int[]{n - 1 - row, col};  // Return (row, col) in terms of board indices
    }
}

File: amazon/909_snakes_and_ladders.py


class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        n = len(board)

        def getPosition(square: int) -> Tuple[int, int]:
            row = (square - 1) // n
            col = (square - 1) % n
            if row % 2 == 1: #reverse column index on odd rows
                col = n -1 - col
            return (n-1-row, col)

        visited = [False] * (n * n + 1)
        queue  = deque([(1,0)])
        visited[1] = True

        while queue:
            curr, moves = queue.popleft()
            
            for i in range (1,7):
                next_square = curr + i
                if next_square > n * n:
                    break

                r,c = getPosition(next_square)

                if board[r][c] != -1:
                    next_square = board[r][c]

                if next_square == n * n: 
                    return moves + 1

                if not visited[next_square]:
                    visited[next_square] = True
                    queue.append((next_square, moves + 1))

        return -1